src/mono/System.Private.CoreLib/src/System/Threading/LowLevelLifoAsyncWaitSemaphore.Browser.Threads.Mono.cs

```diff
@@ -0,0 +1,226 @@
+            if (countsBeforeUpdate == counts)
+            {
+                return counts.SignalCount != 0 || WaitForSignal(timeoutMs);
+            }
+
+            counts = countsBeforeUpdate;
+        }
+#endif
+    }
+
+    private void PrepareAsyncWaitForSignal(int timeoutMs, Action<LowLevelLifoAsyncWaitSemaphore, object?> onSuccess, Action<LowLevelLifoAsyncWaitSemaphore, object?> onTimeout, object? state)
+    {
+        Debug.Assert(timeoutMs > 0 || timeoutMs == -1);
+
+        _onWait();
+
+        PrepareAsyncWaitCore(timeoutMs, new WaitEntry(this, timeoutMs, onSuccess, onTimeout, state));
+        // on success calls InternalAsyncWaitSuccess, on timeout calls InternalAsyncWaitTimeout
+    }
+
+    private static void InternalAsyncWaitTimeout(LowLevelLifoAsyncWaitSemaphore self, WaitEntry internalWaitEntry)
+    {
+        WaitEntry we = internalWaitEntry!;
+        // Unregister the waiter. The wait subsystem used above guarantees that a thread that wakes due to a timeout does
+        // not observe a signal to the object being waited upon.
+        self._separated._counts.InterlockedDecrementWaiterCount();
+        we.OnTimeout(self, we.State);
+    }
+
+    private static void InternalAsyncWaitSuccess(LowLevelLifoAsyncWaitSemaphore self, WaitEntry internalWaitEntry)
+    {
+        WaitEntry we = internalWaitEntry!;
+        // Unregister the waiter if this thread will not be waiting anymore, and try to acquire the semaphore
+        Counts counts = self._separated._counts;
+        while (true)
+        {
+            Debug.Assert(counts.WaiterCount != 0);
+            Counts newCounts = counts;
+            if (counts.SignalCount != 0)
+            {
+                newCounts.DecrementSignalCount();
+                newCounts.DecrementWaiterCount();
+            }
+
+            // This waiter has woken up and this needs to be reflected in the count of waiters signaled to wake
+            if (counts.CountOfWaitersSignaledToWake != 0)
+            {
+                newCounts.DecrementCountOfWaitersSignaledToWake();
+            }
+
+            Counts countsBeforeUpdate = self._separated._counts.InterlockedCompareExchange(newCounts, counts);
+            if (countsBeforeUpdate == counts)
+            {
+                if (counts.SignalCount != 0)
+                {
+                    we.OnSuccess(self, we.State);
+                    return;
+                }
+                break;
+            }
+
+            counts = countsBeforeUpdate;
+        }
+        // if we get here, we need to keep waiting because the SignalCount above was 0 after we did
+        // the CompareExchange - someone took the signal before us.
+        // FIXME(ak): why is the timeoutMs the same as before? wouldn't we starve? why does LowLevelLifoSemaphore.WaitForSignal not decrement timeoutMs?
+        self.PrepareAsyncWaitCore (we.TimeoutMs, we);
```

Comment:
yeah the timeoutms should be reduced, I think.

Reply:
@kouvel are you familiar with this part of the (normal) LowLevelLifoSemaphore code: 

https://github.com/dotnet/runtime/blob/44ea019bbbaab1a0bb2877e6510036456a6495eb/src/libraries/System.Private.CoreLib/src/System/Threading/LowLevelLifoSemaphore.cs#L202-L204

The second time through the outer `while (true)`, if we lost the `InterlockedCompareExchange`, shouldn't `timeoutMs` be decreased by how long we have waited already?

Is it not done that way because measuring the elapsed waiting time is actually too expensive in the common case? (That could certainly be true on wasm, @kg since all the time stuff boils down to an external JS call to `performance.now`)

Reply:
You're right, the timeout should be reduced based on how long the wait took and looks like it's missing in the current implementation. I figure the overhead of timing would be much less than the overhead of WaitCore, though not sure about wasm.

Reply:
performance.now is a little expensive, but it's still worth doing this right i think.

Reply:
ok, I'll update the normal LowLevelLifoSemaphore and the async one to decrement `TimeoutMs`